core/src/consensus.rs

```diff
@@ -47,29 +50,52 @@ pub enum SwitchForkDecision {
 impl SwitchForkDecision {
     pub fn to_vote_instruction(
         &self,
-        vote: Vote,
+        vote: Box<dyn VoteTransaction>,
         vote_account_pubkey: &Pubkey,
         authorized_voter_pubkey: &Pubkey,
     ) -> Option<Instruction> {
-        match self {
-            SwitchForkDecision::FailedSwitchThreshold(_, total_stake) => {
+        match (
+            self,
+            vote.as_any().downcast_ref::<Vote>(),
+            vote.as_any().downcast_ref::<VoteStateUpdate>(),
+        ) {
+            (SwitchForkDecision::FailedSwitchThreshold(_, total_stake), _, _) => {
                 assert_ne!(*total_stake, 0);
                 None
             }
-            SwitchForkDecision::FailedSwitchDuplicateRollback(_) => None,
-            SwitchForkDecision::SameFork => Some(vote_instruction::vote(
+            (SwitchForkDecision::FailedSwitchDuplicateRollback(_), _, _) => None,
+            (SwitchForkDecision::SameFork, Some(v), None) => Some(vote_instruction::vote(
                 vote_account_pubkey,
                 authorized_voter_pubkey,
-                vote,
+                v.clone(),
             )),
-            SwitchForkDecision::SwitchProof(switch_proof_hash) => {
+            (SwitchForkDecision::SameFork, None, Some(v)) => {
+                Some(vote_instruction::update_vote_state(
+                    vote_account_pubkey,
+                    authorized_voter_pubkey,
+                    v.clone(),
+                ))
+            }
+            (SwitchForkDecision::SwitchProof(switch_proof_hash), Some(v), None) => {
                 Some(vote_instruction::vote_switch(
                     vote_account_pubkey,
                     authorized_voter_pubkey,
-                    vote,
+                    v.clone(),
                     *switch_proof_hash,
                 ))
             }
+            // TODO: do we need a separate instruction here?
```

Comment:
ick yeah good catch, eventually for slashing we'll need to be able to distinguish between votes with a switch proof vs without, like `vote` vs `vote_switch` above. Might be good to add in that separate instruction now